### PR TITLE
Added encryption key metadata to send additional information about th…

### PR DESCRIPTION
**Pull request by saandrews**
Addresses issue discussed in #797

### CODE REVIEW

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/MessageCrypto.java [hunk 1]
```diff
@@ -118,7 +120,7 @@
 
     public MessageCrypto(boolean keyGenNeeded) {
 
-        encryptedDataKeyMap = new ConcurrentHashMap<String, byte[]>();
+        encryptedDataKeyMap = new ConcurrentHashMap<String, Pair<byte[], List<KeyValue>>>();
```
Comment (rdhabalia) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/MessageCrypto.java]: if we already have `EncryptionKeyInfo` which contains `byte[]` and `List<KeyValue>>` then can we use the same instead `Pair`?

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/MessageCrypto.java [hunk 2]
```diff
@@ -356,22 +358,37 @@ public boolean removeKeyCipher(String keyName) {
...
         if (encKeys.isEmpty()) {
             return payload;
         }
 
         // Update message metadata with encrypted data key
-        encKeys.forEach(keyName -> {
-            if (encryptedDataKeyMap.get(keyName) != null) {
-                msgMetadata.addEncryptionKeys(KeyByteValue.newBuilder().setKey(keyName)
-                    .setValue(ByteString.copyFrom(encryptedDataKeyMap.get(keyName))).build());
+        List<String> keyNameList = encKeys.values();
```
Comment (rdhabalia) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/MessageCrypto.java]: `encKeys.values()` creates a new list..instead we can use `forEach()` ?

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/MessageCrypto.java [hunk 3]
```diff
@@ -356,22 +358,37 @@ public boolean removeKeyCipher(String keyName) {
...
+                            .setValue(ByteString.copyFrom(keyInfo.getLeft())).addAllMetadata(keyInfo.getRight()).build());
+                }
             } else {
-                log.warn("Failed to find encrypted Data key for key {}.", keyName);
+                // We should never reach here.
+                log.error("Failed to find encrypted Data key for key {}.", keyName);
             }
-        });
+
+        }
```
Comment (rdhabalia) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/MessageCrypto.java]: can we remove `//TODO` at line 393 if it's already implemented?

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/MessageCrypto.java [hunk 4]
```diff
@@ -511,7 +529,7 @@ private ByteBuf decryptDataKeyAndData(MessageMetadata msgMetadata, ByteBuf paylo
                     break;
                 }
             } catch (Exception e) {
-                // First time, entry won't be present in cache 
+                // First time, entry won't be present in cache
```
Comment (rdhabalia) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/MessageCrypto.java]: can we add debug log here?

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/MessageCrypto.java [hunk 5]
```diff
@@ -487,14 +505,14 @@ private ByteBuf decryptData(SecretKey dataKeySecret, MessageMetadata msgMetadata
...
+    private ByteBuf getKeyAndDecryptData(MessageMetadata msgMetadata, ByteBuf payload) {
 
         ByteBuf decryptedData = null;
 
-        List<KeyByteValue> encKeys = msgMetadata.getEncryptionKeysList();
+        List<EncryptionKeys> encKeys = msgMetadata.getEncryptionKeysList();
 
         // Go through all keys to retrieve data key from cache
-        for(int i=0; i<encKeys.size(); i++) {
+        for (int i = 0; i < encKeys.size(); i++) {
```
Comment (rdhabalia) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/MessageCrypto.java]: `for(KeyByteValue encKey: encKeys) {..} ..?`

**Comment by rdhabalia** [comment]
@saandrews 
I can see few encryption related test-failures at [jenkins](https://builds.apache.org/job/pulsar-pull-request/469/)
```
org.apache.pulsar.client.api.SimpleProducerConsumerTest.testECDSAEncryption
org.apache.pulsar.client.api.SimpleProducerConsumerTest.testEncryptionFailure
org.apache.pulsar.client.api.SimpleProducerConsumerTest.testRSAEncryption
```

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/MessageCrypto.java (hunk 1 above)
Comment (saandrews) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/MessageCrypto.java]: Fixed. The one difference here is that this map stores encrypted key vs the KeyReader returns actual key.

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/MessageCrypto.java (hunk 2 above)
Comment (saandrews) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/MessageCrypto.java]: Its done to propagate exception above

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/MessageCrypto.java (hunk 3 above)
Comment (saandrews) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/MessageCrypto.java]: This part is not done yet. Not critical for initial release. But we should revisit.

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/MessageCrypto.java (hunk 5 above)
Comment (saandrews) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/MessageCrypto.java]: It needs to set decryptedData which can't be done from the lambda

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/MessageCrypto.java (hunk 4 above)
Comment (saandrews) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/MessageCrypto.java]: Added

**Comment by saandrews** [comment]
retest this please

**Comment by saandrews** [comment]
retest this please

File: pulsar-client/src/main/java/org/apache/pulsar/client/api/CryptoKeyReader.java [hunk 6]
```diff
@@ -18,27 +18,36 @@
  */
 package org.apache.pulsar.client.api;
 
+import java.util.List;
+
+import org.apache.pulsar.client.impl.EncryptionKeyInfo;
```
Comment (merlimat) [inline pulsar-client/src/main/java/org/apache/pulsar/client/api/CryptoKeyReader.java]: If `EncryptionKeyInfo` is part of public API it needs to be in `org.apache.pulsar.client.api` package

File: pulsar-client/src/main/java/org/apache/pulsar/client/api/CryptoKeyReader.java [hunk 7]
```diff
@@ -18,27 +18,36 @@
...
+import org.apache.pulsar.common.api.proto.PulsarApi.KeyValue;
```
Comment (merlimat) [inline pulsar-client/src/main/java/org/apache/pulsar/client/api/CryptoKeyReader.java]: Protobuf types must not be in public api. Use a `Map<String, String>`

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/MessageCrypto.java [hunk 8]
```diff
@@ -511,7 +529,8 @@ private ByteBuf decryptDataKeyAndData(MessageMetadata msgMetadata, ByteBuf paylo
                     break;
                 }
             } catch (Exception e) {
-                // First time, entry won't be present in cache 
+                // First time, entry won't be present in cache
+                log.debug("Failed to decrypt data or data key is not in cache. Will attempt to refresh");
```
Comment (merlimat) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/MessageCrypto.java]: Please add more contextual info (topic, etc..)

Comment (merlimat) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/MessageCrypto.java]: Exception..

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/MessageCrypto.java [hunk 9]
```diff
@@ -511,7 +529,8 @@ private ByteBuf decryptDataKeyAndData(MessageMetadata msgMetadata, ByteBuf paylo
                     break;
                 }
             } catch (Exception e) {
-                // First time, entry won't be present in cache 
+                // First time, entry won't be present in cache
```
Comment (merlimat) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/MessageCrypto.java]: Why should we get an exception if it's not in cache? Exceptions are expensive, should we just check for null before?

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/MessageCrypto.java (hunk 8 above)
Comment (saandrews) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/MessageCrypto.java]: Added logCtx, so we can log topic/producername/id for producer and topic/subscription for consumer.

**Pull request merged**